src/Raven.Server/NotificationCenter/SlowWriteNotifications.cs

```diff
@@ -0,0 +1,135 @@
+﻿using System;
+using System.Threading;
+using Raven.Client.Documents.Conventions;
+using Raven.Client.Documents.Session;
+using Raven.Client.Util;
+using Raven.Server.NotificationCenter.Notifications;
+using Raven.Server.NotificationCenter.Notifications.Details;
+using Sparrow.Collections.LockFree;
+using Sparrow.Json;
+
+namespace Raven.Server.NotificationCenter
+{
+    public class SlowWriteNotifications : IDisposable
+    {
+        internal TimeSpan UpdateFrequency = TimeSpan.FromSeconds(15);
+        private readonly NotificationCenter _notificationCenter;
+        private readonly NotificationsStorage _notificationsStorage;
+        private readonly string _database;
+        private Timer _timer;
+        private bool _updateNotificationInStorageRequired;
+        private readonly object _pagerCreationLock = new object();
+        private readonly ConcurrentDictionary<string, SlowWritesDetails.SlowWriteInfo> _slowWrites;
+
+        public SlowWriteNotifications(NotificationCenter notificationCenter, NotificationsStorage notificationsStorage, string database)
+        {
+            _notificationCenter = notificationCenter;
+            _notificationsStorage = notificationsStorage;
+            _database = database;
+            _slowWrites = new ConcurrentDictionary<string, SlowWritesDetails.SlowWriteInfo>();
+        }
+
+        public void Add(string path, double dataSizeInMb, double durationInSec)
+        {
+            var now = SystemTime.UtcNow;
+
+            if (_slowWrites.TryGetValue(path, out var info))
+            {
+                if (now - info.Date < UpdateFrequency)
+                    return;
+
+                info.DataWrittenInMb = dataSizeInMb;
+                info.DurationInSec = durationInSec;
+                info.SpeedInMbPerSec = dataSizeInMb / durationInSec;
+                info.Date = now;
+            }
+            else
+            {
+                info = new SlowWritesDetails.SlowWriteInfo
+                {
+                    Path = path,
+                    DataWrittenInMb = dataSizeInMb,
+                    DurationInSec = durationInSec,
+                    SpeedInMbPerSec = dataSizeInMb / durationInSec,
+                    Date = now
+                };
+            }
+
+            _slowWrites.AddOrUpdate(path, info, (s, i) => info);
```

Comment:
We already did the check, why are we calling add or update here?

Reply:
Right, we can call `TryAdd` if key isn't there. If it is there we are updating in-memory reference anyway - no need to call `AddOrUpdate`. Thanks